### PR TITLE
Added mapReduce on Model

### PR DESCRIPTION
**Pull request by RagibHasin**
Sorry for late. I was a little busy with my classes.

Yes, it went difficult to take Instance type as a generic parameter and I got stuck. So I've worked it out with a decorator on Instance. Sending the PR.

### CODE REVIEW

File: test/MapReduce.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                out: { 
+                    replace: "mapReduced" 
+                }, query: { status: "A" } }).then(() => {
+                let reducedModel = new Iridium.Model<Iridium.MapReducedDocument<string, number>, MapReducedInstance>(core, MapReducedInstance);
+                return reducedModel.find().toArray()
+            })
+            return chai.expect(t).to.eventually.exist.and.have.length(2);
+        });
+
+        it("should correctly map and reduce inline", () => {
```
Comment (SPARTAN563) [inline test/MapReduce.ts]: I'd also love to see a test case that ensures it throws an error and doesn't do anything when you specify `out: "something"`.

File: lib/Model.ts [hunk 2]
```diff
@@ -692,6 +693,52 @@ export class Model<TDocument extends { _id?: any }, TInstance> {
         });
     }
 
+    private _mapReduce<Key, Value>(instanceType: InstanceImplementation<MapReducedDocument<Key, Value>, any>, options: MongoDB.MapReduceOptions): Bluebird<void> {
```
Comment (SPARTAN563) [inline lib/Model.ts]: Looking at the implementations, you may be able to write things such that you don't need to branch calls to `this.collection.mapReduce()` (by extracting the `mapReduceOptions` first and checking whether `options.out` is `{ inline: 1 }` in the callback before `resolve(data)`).

File: package.json [hunk 3]
```diff
@@ -22,7 +22,7 @@
     "build": "tsc",
     "pretest": "npm run build",
     "prepublish": "npm run build",
-    "test": "mocha --opts test/mocha.opts dist/test",
+    "test": "mocha --opts test/mocha.opts dist/test/*.js && mocha --opts test/mocha.opts dist/test",
```
Comment (SPARTAN563) [inline package.json]: Why was this necessary? Was something not working correctly on your side?

File: test/MapReduce.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                }, reduce: function (k: string, v: number[]) {
+                    var sum = 0;
+                    v.forEach((val, i, arr) => { sum += val });
+                    return sum;
+                }
+            }, { out: { inline: 1 }, query: { status: "A" } })
+            return chai.expect(t).to.eventually.exist.and.have.length(2);
+        });
+
+        /*
```
Comment (SPARTAN563) [inline test/MapReduce.ts]: Do you mind removing this? It's obviously nice as a reference point, but that code is already elsewhere in the codebase.

**Comment by SPARTAN563** [comment]
Oh, also, if you wouldn't mind quickly rebasing on `master`, it should fix the low code coverage you're seeing (that was my mistake and I only noticed it now).

File: lib/Model.ts [hunk 5]
```diff
@@ -692,6 +693,52 @@ export class Model<TDocument extends { _id?: any }, TInstance> {
...
+            else
+                this.collection.mapReduce(instanceType.mapReduceOptions.map, instanceType.mapReduceOptions.reduce, options, (err, data) => {
+                    if (err) return reject(err)
+                    return resolve()
+                })
+        })
+    }
+
+    /**
+     *
```
Comment (RagibHasin) [inline lib/Model.ts]: In the inline version where the first parameter is an object with map and reduce function, it can work without any `option.out.inline = 1` being passed, cause it treats it as inline by default. If `option` is constrained as you described, this default behaviour won't be possible.

File: test/MapReduce.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                { cust_id: "A123", amount: 250, status: "A" },
+                { cust_id: "A123", amount: 300, status: "B" },
+                { cust_id: "B212", amount: 200, status: "A" }
+            ]));
+        });
+
+        it("should correctly map and reduce with model", () => {
+            let t = model.mapReduce(MapReducedInstance, { 
+                out: { 
+                    replace: "mapReduced"
```
Comment (RagibHasin) [inline test/MapReduce.ts]: Yes, I also think so. I'll change the type of the `options.out` field to an enum.

File: test/MapReduce.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            ]));
+        });
+
+        it("should correctly map and reduce with model", () => {
+            let t = model.mapReduce(MapReducedInstance, { out:{replace:'mapReduced'}, query: { status: "A" } }).then(()=>{
+                let reducedModel = new Iridium.Model<Iridium.MapReducedDocument<string, number>, MapReducedInstance>(core, MapReducedInstance);
+                return reducedModel.find().toArray()
+            })
+            t.then(instances => {
+                instances.forEach((val, i, arr) => { console.log(JSON.stringify(val)) });
```
Comment (RagibHasin) [inline test/MapReduce.ts]: Yes, 😅. Forgot to remove that before commit. Done now.

File: test/MapReduce.ts (hunk 1 above)
Comment (RagibHasin) [inline test/MapReduce.ts]: Good idea. I'm on it.

File: package.json (hunk 3 above)
Comment (RagibHasin) [inline package.json]: Yes. There was no recursive option set in `test/mocha.opts` and I am not so keen to Mocha. I'll remove this.

**Comment by RagibHasin** [comment]
Thanks, that worked. I actually got afraid where I made the mistake.

File: lib/Model.ts (hunk 2 above)
Comment (RagibHasin) [inline lib/Model.ts]: Yes, I got the point. Moved the private function into the public functions branch.

File: test/MapReduce.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
+import * as Iridium from "../iridium";
```
Comment (RagibHasin) [inline test/MapReduce.ts]: Welcome.

**Comment by RagibHasin** [comment]
A hope now it's OK to merge.

File: lib/MapReduce.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+ * 
+ * Extends `MongoDB.MapReduceOptions` with additional map and reduce field.
+ */
+export interface MapReduceFunctions<TDocument, Key, Value> {
+    map: MapFunction<TDocument>
+    reduce: ReduceFunction<Key, Value>
+}
+
+export interface MapReduceOptions extends MongoDB.MapReduceOptions {
+    out?: "inline" | "replace" | "merge" | "reduce"
```
Comment (SPARTAN563) [inline lib/MapReduce.ts]: Very nice idea to merge with the `@Iridium.Collection` decorator.

File: test/MapReduce.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+import * as MongoDB from "mongodb";
+import { Cursor } from "../lib/Cursor";
+import * as Promise from "bluebird";
+import * as _ from "lodash";
+import * as chai from "chai";
+
+// This test folows the example depicted on https://docs.mongodb.com/manual/core/map-reduce/
```
Comment (SPARTAN563) [inline test/MapReduce.ts]: Nice :+1:

File: test/MapReduce.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+class MapReducedInstance extends Iridium.Instance<Iridium.MapReducedDocument<string, number>, MapReducedInstance>{
+    static collection = "mapReduced";
+    static mapReduceOptions = {
+        map: function (this: TestDocument) {
+            emit(this.cust_id, this.amount);
+        },
+        reduce: function (key: string, values: number[]) {
+            return values.reduce((sum, val) => sum + val, 0)
```
Comment (SPARTAN563) [inline test/MapReduce.ts]: Looks clean :+1:

File: test/MapReduce.ts [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+                { cust_id: "A123", amount: 300, status: "B" },
+                { cust_id: "B212", amount: 200, status: "A" }
+            ]));
+        });
+
+        it("should correctly map and reduce with model", () => {
+            let t = model.mapReduce(MapReducedInstance, {
+                out: "replace", query: { status: "A" }
+            }).then(() => {
+                let reducedModel = new Iridium.Model<Iridium.MapReducedDocument<string, number>, MapReducedInstance>(core, MapReducedInstance);
```
Comment (SPARTAN563) [inline test/MapReduce.ts]: Maybe define the model before the test and run `reducedModel.remove()` to ensure it's empty to start. I can imagine a case where you've previously had the test work correctly and there are two documents in the collection, the `mapReduce` does nothing but this test passes as a result.

**Comment by SPARTAN563** [comment]
Looks really good, awesome work @RagibHasin and let me know if you want to make changes to the test, otherwise I'll go ahead and merge this for the `v7.2.x` release cycle.

File: test/MapReduce.ts (hunk 12 above)
Comment (RagibHasin) [inline test/MapReduce.ts]: Yes, it'll be good to fix it there. Thanks.

File: test/MapReduce.ts [hunk 13]
```diff
@@ -61,12 +61,11 @@ describe("Model", () => {
         });
 
         it("should correctly map and reduce with model", () => {
+            let reducedModel = new Iridium.Model<Iridium.MapReducedDocument<string, number>, MapReducedInstance>(core, MapReducedInstance);
+            reducedModel.remove()
```
Comment (SPARTAN563) [inline test/MapReduce.ts]: You're gonna need to wait on this, so `return reducedModel.remove().then() => /* your map reduce test */`

Comment (SPARTAN563) [inline test/MapReduce.ts]: Sorry if I wasn't clear initially

Comment (RagibHasin) [inline test/MapReduce.ts]: Oh, missed that. Sorry.

**Comment by SPARTAN563** [comment]
Looks prefect, gonna wait for the tests to complete (they should happily pass) and then I'll merge this. Thanks for the PR :+1:

**Comment by RagibHasin** [comment]
Welcome.
